### ISSUE TITLE
com.amazonaws.internal.keyvaluestore.KeyNotFoundException

### ISSUE THREAD

**Issue by zmeid** [aws-amplify/aws-sdk-android#1392]
**Describe the bug**
I was able to get tokens from AWSMobileClient successfully. But, 2 days ago i started get the error: "getTokens does not support retrieving tokens while signed-out". It was weird because i was trying to get tokens right after logging the user in, in another activity which is user profile. So, i was sure that user was not in signout state When i checked other logs, i saw that AWSMobileClient was having trouble with KeyValueStore. I am getting KeyNotFoundException when i try to get tokens. I read other issues related KeyValueStore in here. Here is what i tried so far:

-Clearing cache and data.
-Uninstalling the app.
-Restarting the device.
-Logging in with different credentials.
-Registering a new user and logging with the new user.
-Downgrading AWSMobileClient to 2.15.2.

No luck so far.

**To Reproduce**
Here is how i login:
```
    fun signInUser(username: String, password: String, cognitoSignInResponse: CognitoSignInResponse) {
        user = cognitoUserPool.getUser(username)

        val authenticationHandler = object : AuthenticationHandler {
            override fun authenticationChallenge(continuation: ChallengeContinuation?) {}

            override fun getMFACode(continuation: MultiFactorAuthenticationContinuation?) {}

            override fun onSuccess(userSession: CognitoUserSession?, newDevice: CognitoDevice?) {
                cognitoSignInResponse.onSuccess(userSession)
            }

            override fun getAuthenticationDetails(authenticationContinuation: AuthenticationContinuation,
                                                  serId: String) {
                // The API needs user sign-in credentials to continue
                val authenticationDetails = AuthenticationDetails(username, password, null)
                // Pass the user sign-in credentials to the continuation
                authenticationContinuation.setAuthenticationDetails(authenticationDetails)
                // Allow the sign-in to continue
                authenticationContinuation.continueTask()
            }

            override fun onFailure(exception: Exception) {
                cognitoSignInResponse.onFailure(exception)
            }
        }
        user.getSessionInBackground(authenticationHandler)
    }
```
Here is how i try to get tokens:
```
        AWSMobileClient.getInstance().initialize(context, awsConfig, object : Callback<UserStateDetails> {
            override fun onResult(result: UserStateDetails?) {

                AWSMobileClient.getInstance().getTokens(object : Callback<Tokens> {
                    override fun onResult(result: Tokens?) {
                        accessTokenCallback.onSuccess(result?.accessToken?.tokenString)
                    }

                    override fun onError(e: java.lang.Exception?) {
                        Log.e("Cognito error:", " " + e?.toString())
                        accessTokenCallback.onFailure(e)
                    }
                })
            }

            override fun onError(e: java.lang.Exception?) {
                Log.e("Cognito error:", " " + e?.toString())
                accessTokenCallback.onFailure(e)
            }
        })
```

**Which AWS service(s) are affected?**
AWSMobileClient

**Expected behavior**
Getting cached tokens

**Environment Information (please complete the following information):**
 - AWS Android SDK Version: com.amazonaws:aws-android-sdk-mobile-client:2.16.5
 - Device: Samsung S8(Android 9) and Virtual Device Pixel 2(API 26)
 - Specific to simulators: No

**Error log:**
```
2019-12-24 15:08:33.468 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.487 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = provider
2019-12-24 15:08:33.490 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.497 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = token
2019-12-24 15:08:33.499 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.502 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = cognitoIdentityId
2019-12-24 15:08:33.504 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.510 29867-32338/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = isFederationEnabled
2019-12-24 15:08:33.515 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.519 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = provider
2019-12-24 15:08:33.520 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.526 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = token
2019-12-24 15:08:33.528 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.533 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = provider
2019-12-24 15:08:33.535 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.543 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = token
2019-12-24 15:08:33.548 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.565 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = cognitoIdentityId
2019-12-24 15:08:33.569 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: com.amazonaws.internal.keyvaluestore.KeyNotFoundException: Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.mobile.client.aesKeyStoreAlias
2019-12-24 15:08:33.579 29867-32343/com.xxx.app.debug E/AWSKeyValueStore: Error in retrieving the decryption key used to decrypt the data from the persistent store. Returning null for the requested dataKey = isFederationEnabled
2019-12-24 15:08:33.582 29867-32343/com.xxx.app.debug E/Cognito error::  java.lang.Exception: getTokens does not support retrieving tokens while signed-out
```

**Comment by TrekSoft** [comment]
Hey, in the past issues related to this were due to problems in the configuration file - could you check these two issues and see if your issue is related: #948 and #942

**Comment by zmeid** [comment]
I checked the issues which you mentioned. My awsconfiguration.json is configured correctly. The issues are not related. Here is template of my awsconfiguration.json:

`{
  "Version": "1.0",
  "CognitoUserPool": {
    "Default": {
      "AppClientId": "xxxxxxxxxxxxxx",
      "PoolId": "xxxxx_xxxxxxxx",
      "Region": "xxxxx"
    }
  }
}`

**Comment by zmeid** [comment]
I noticed that "CognitoIdentityProviderCache.aesKeyStoreAlias" can be retrieved successfully while "com.amazonaws.mobile.client.aesKeyStoreAlias" cannot. 

Since i am stuck with AwsMobileClient, i am going to use cognitoUser.getSessionInBackground() to get user tokens. Looks like getSessionInBackground() works. Any other advices?

**Comment by dvphu79** [comment]
I run Android emulator and see those errors too.

I find out Android emulator have no internet access, so I set Google DNS 8.8.8.8 in my Mac and restart Android emulator.

**Comment by TrekSoft** [comment]
So just to be clear, are you unblocked now? Or are you still having the original problem?

**Comment by dvphu79** [comment]
yes, I am unblocked now

**Comment by TrekSoft** [comment]
Awesome - apologies for the delay

**Comment by vinod2126** [comment]
Error AWSKeyValueStore  -->  com.amazonaws.internal.keyvaluestore.KeyNotFoundException  Error occurred while accessing AndroidKeyStore to retrieve the key for keyAlias: com.amazonaws.android.auth.aesKeyStoreAlias


error in android automation pipeline ...using device farm

**Comment by tylerjroach** [comment]
@vinod2126 It looks like the customer was able to resolve the original issue and it was solved by simply restarting the device/computer. If you are still experiencing this isuse, can you open a new issue providing information on the Amplify version, device info, and if it is happening on all your devices running in Device Farm, or just a single device?
